fix(ngMock): prevent memory leak due to data attached to `$rootElement`

gkalpak opened the pull request:
Starting with 88bb551, `ngMock` will attach the `$injector` to the `$rootElement`, but will never clean it up, resulting in a memory leak. Since a new `$rootElement` is created for every test, this leak causes Karma to crash on large test-suites.
The problem was not detected by our internal tests, because we do our own clean-up in
`testabilityPatch.js`.

This commit prevents the memory leak, by cleaning up all data attached to `$rootElement` after each test.

Fixes #14094

gkalpak committed:
fix(ngMock): prevent memory leak due to data attached to `$rootElement`

Starting with 88bb551, `ngMock` will attach the `$injector` to the `$rootElement`, but will never
clean it up, resulting in a memory leak. Since a new `$rootElement` is created for every test,
this leak causes Karma to crash on large test-suites.
The problem was not detected by our internal tests, because we do our own clean-up in
`testabilityPatch.js`.

This commit prevents the memory leak, by cleaning up all data attached to `$rootElement` after each
test.

Fixes #14094

petebacondarwin commented on test/ngMock/angular-mocksSpec.js:
```diff
@@ -2408,9 +2408,134 @@ describe('ngMockE2E', function() {
 });
+
+
+describe('don\'t leak memory between tests', function() {
```
I prefer describe block descriptions to be nouns of things that are being tested...

But I see you are following the pattern above :-(

gkalpak replied:
I don't like the descriptions either. I'll tweak them.

petebacondarwin commented:
So the idea of the tests is that you have to run a test beforehand that sets up the spies so that you can check to see that the root element was in fact cleaned up?
I think that the wording of the describe and it blocks could be tweaked so that we remember what they are doing in 12 months time.

petebacondarwin commented on test/ngMock/angular-mocksSpec.js:
```diff
@@ -2408,9 +2408,134 @@ describe('ngMockE2E', function() {
+  });
+});
+
+describe('don\'t leak memory between tests with mocked `$rootScope`', function() {
```
should this be "with mocked `$rootElement`"?

gkalpak replied:
Yup :)

petebacondarwin commented on test/ngMock/angular-mocksSpec.js:
```diff
@@ -2408,9 +2408,134 @@ describe('ngMockE2E', function() {
+      expect(cleanUpElems.length).toBe(1);
+      expect(cleanUpElems[0][0]).toBe(prevRootElement[0]);
+    }
+  });
```
Do we need to do anything to tidy up the spies that we are holding on to?

gkalpak replied:
Theoretically, no. Jasmine will restore the original objects after each test. We hold on to the spies, but the original object is restored. When the `define` blocks are done, the references to those spies are useless (and should be garbage collected), but don't affect the environment in any way.

jbedard commented on src/ngMock/angular-mocks.js:
```diff
@@ -2600,7 +2602,17 @@ if (window.jasmine || window.mocha) {
+        if (elem && elem.off) elem.off();
+        if (elem && elem.removeData) elem.removeData();
+      });
+      if (window.jQuery) window.jQuery.cleanData(cleanUpElems);
```
I think this can be `jqLite.cleanData(cleanUpElements)`. Then you might not need the `off` or `removeData` calls above?

petebacondarwin replied:
I was wondering about that too.

gkalpak replied:
I copied this over from `testabilityPatch.js`, but I think you are right.

From a quick look, I alos think there might be a leak here (and in `testabilityPatch.js`):
Passing an array of jqLite/jQuery elements won't properly clean up. `cleanData` expects an array of "raw" node elements.

I'll try this out.

jbedard replied:
The `jqLite` version of `cleanData` was added [fairly recently](https://github.com/angular/angular.js/commit/96288d02d316ff1508b2e0673f34291f9abd1c7f#diff-8e0d9e5c18fe4f1050484080525d4367R192)...

gkalpak replied:
Yeah, that would explain it :)

gkalpak committed:
fixup 1

- Group into a `describe` block
- Tweak test descriptions
- Fix typos

gkalpak commented:
I has to rebase this on top of master (because the commit this is supposed to fix was reverted).

I incorporated the original commit (attaching `$injector` to `$rootScope`) in this PR's 1st commit. (Other than that, I haven't done any other changes in the 1st commit, so you don't have to re-review the other parts.)

I added a 2nd commit, that addressed your feedback (except for https://github.com/angular/angular.js/pull/14098#discussion_r53570545 - I'll play with that next).
I tried to make the test descriptions as clear as possible and add comments explaining parts of the test logic that felt unclear, but I'm open to suggestions for better explanations :smiley:

gkalpak committed:
fixup 2

- Unify the clean-up paths for tests with/without jQuery, using `angular.element.cleanData()`.
- Remove redundant calls to `.off().removeData()`.
- Fix bug in `testabilityPatch.js`, where `cleanData()` was called with an array of jqLite/jQuery
  elements (instead of "raw" nodes).

gkalpak commented:
I added a 3rd commit that uses `jqLite.cleanData(...)`.

gkalpak committed:
fixup 3

- Reset `originalRootElement` for each test to avoid.
- Fix clean-up break, if `$rootElement` was never initialized.
- Fix clean-up break, if decorated `$rootElement` was falsy (e.g. `null`).
- (Rename `cleanUpElems` to `cleanUpNodes` - since it's "raw" nodes, not jq-wrapped elements.)

gkalpak commented:
And a 4th commit (fixup 3) that fixes a couple of corner cases.
<sub>(BTW, the commit messages contain more details about the contents of each fixup commit.)</sub>

petebacondarwin commented:
LGTM - please squash and merge.

gkalpak commented:
Backported to `v1.5.x` (75373dd) and `v1.4.x` (571e323).
